be/src/exec/vectorized/sorting/sort_column.cpp

```diff
@@ -97,6 +111,229 @@ class ColumnSorter final : public ColumnVisitorAdapter<ColumnSorter> {
+                                            _build_tie, _limit, &_limited));
+        _prune_limit();
+        return Status::OK();
+    }
+
+    template <typename T>
+    Status do_visit(const vectorized::ObjectColumn<T>& column) {
+        DCHECK(false) << "not supported";
+        return Status::NotSupported("TOOD");
+    }
+
+    Status do_visit(const vectorized::JsonColumn& column) {
+        auto cmp = [&](const PermutationItem& lhs, const PermutationItem& rhs) {
+            auto& lhs_col = _vertical_columns[lhs.chunk_index];
+            auto& rhs_col = _vertical_columns[rhs.chunk_index];
+            return lhs_col->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *rhs_col, _is_null_first ? -1 : 1);
+        };
+
+        RETURN_IF_ERROR(sort_and_tie_helper(_cancel, &column, _is_asc_order, _permutation, _tie, cmp, _range,
+                                            _build_tie, _limit, &_limited));
+        _prune_limit();
+        return Status::OK();
+    }
+
+private:
+    template <class T>
+    struct InlineChunkItem {
+        uint32_t chunk_index;
+        uint32_t index_in_chunk;
+
+        T inline_value;
+    };
+
+    template <class T>
+    using InlinedChunkPermutation = std::vector<InlineChunkItem<T>>;
+
+    bool _need_inline_value() {
+        // TODO: figure out the inflection point
+        // If limit exceeds 1/5 rows, inline will has benefits
+        size_t total_rows = _permutation.size();
+        return _limit > (total_rows / 5);
+    }
+
+    template <class T, class DataContainer>
+    InlinedChunkPermutation<T> _create_inlined_permutation(const std::vector<DataContainer>& containers) {
+        InlinedChunkPermutation<T> result(_permutation.size());
+        for (size_t i = 0; i < _permutation.size(); i++) {
+            int chunk_index = _permutation[i].chunk_index;
+            int index_in_chunk = _permutation[i].index_in_chunk;
+            result[i].chunk_index = chunk_index;
+            result[i].index_in_chunk = index_in_chunk;
+            result[i].inline_value = (*containers[chunk_index])[index_in_chunk];
+        }
+        return result;
+    }
+
+    template <class T>
+    void _restore_inlined_permutation(const InlinedChunkPermutation<T> inlined) {
+        size_t n = std::min(inlined.size(), _limited);
+        for (size_t i = 0; i < n; i++) {
+            _permutation[i].chunk_index = inlined[i].chunk_index;
+            _permutation[i].index_in_chunk = inlined[i].index_in_chunk;
+        }
+    }
+
+    void _prune_limit() {
+        if (_limited < _permutation.size()) {
+            _permutation.resize(_limited);
+            _tie.resize(_limited);
+        }
+    }
+
+    const bool& _cancel;
+    const bool _is_asc_order;
+    const bool _is_null_first;
+
+    std::vector<ColumnPtr> _vertical_columns;
+    Permutation& _permutation;
+    Tie& _tie;
+    std::pair<int, int> _range;
+    bool _build_tie;
+    size_t _limit;
```

Comment:
what's the _limit and _limited difference?

Reply:
In SQL, user could specify `order by xxx limit n`. That's `_limit` means.

But in column-wise sorting, we could not simply apply the limit to the partial sort, but calculate the actual limit as `_limited`.

Consider sorting the column `a`, `b` limit 3. Column `a` consists of `[1, 2, 2, 2, 4]`, even the `limit` is 3, we could not prune the column into `[1, 2, 2]`, since there's multiple `2` in this column, they need to be compared through the column `b`. At this case, we could prune the `_limited` to 4, which means `[1,2,2,2]`.

In implementation of partial sort, we perform the sorting in two stages:
1. Partial sort with limit 3
2. Take the last element `2`, find equal elements in the column and append them to the tail
3. Calculate the `_limited` as limit + equal-elements

Reply:
I see. So I think we need a better name for `_limited`?